### PR TITLE
JDK-8299790: os::print_hex_dump is racy

### PR DESCRIPTION
**Pull request by tstuefe**
Tiny fix for a tiny problem. 

`os::print_hex_dump` uses `os::is_readable_pointer` to check the to-be-printed memory for readability; `os::is_readable_pointer` uses `SafeFetch` to probe the memory for access, which is good, but then, by the time we actually print that information, we reread the memory location again. It may be unreadable now (either because the region had been unmapped or protected by a concurrent thread), and we would crash the VM.

The patch rewrites the function to not use `os::is_readable_pointer`, but to use `SafeFetch` to read from memory directly and then use the result of that read for printing. That requires a bit of bit fiddling, since we only can read word-wise, but the hex-dump could be in units between bytes and qwords.

Tests: manual and GHA-driven gtests on all platforms. The gtests test this function exhaustively.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8299790](https://bugs.openjdk.org/browse/JDK-8299790): os::print_hex_dump is racy (**Bug** - P4)

### Reviewers
 * [Aleksey Shipilev](https://openjdk.org/census#shade) (@shipilev - **Reviewer**) ⚠️ Review applies to [42794c20](https://git.openjdk.org/jdk/pull/14895/files/42794c20af025102e4ba1d45cc2f072d84fc1f2a)
 * [David Holmes](https://openjdk.org/census#dholmes) (@dholmes-ora - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/14895/head:pull/14895` \
`$ git checkout pull/14895`

Update a local copy of the PR: \
`$ git checkout pull/14895` \
`$ git pull https://git.openjdk.org/jdk.git pull/14895/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 14895`

View PR using the GUI difftool: \
`$ git pr show -t 14895`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/14895.diff">https://git.openjdk.org/jdk/pull/14895.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/14895#issuecomment-1637177154)

### CODE REVIEW

File: src/hotspot/share/runtime/os.cpp
Comment (shipilev) [inline src/hotspot/share/runtime/os.cpp:999]: Pull the `??????` printout straight into `print_hex_location` and remove the need for the return value?

File: src/hotspot/share/runtime/os.cpp
Comment (shipilev) [inline src/hotspot/share/runtime/os.cpp:946]: Unused?

File: src/hotspot/share/runtime/os.cpp
Comment (shipilev) [inline src/hotspot/share/runtime/os.cpp:960]: This implies endianness, it is not? We handle endianness for 64-bit version below.

**Review by shipilev — COMMENTED** [review]
Some comments...

File: src/hotspot/share/runtime/os.cpp
Comment (tstuefe) [reply src/hotspot/share/runtime/os.cpp:960]: Sure, for completeness; though I don't think we have a 32-bit big-endian platform.

**Comment by tstuefe** [comment]
Thank you @shipilev. I worked in your feedback and massaged the code a bit.

**Comment by tstuefe** [comment]
Friendly ping... still need a second review.

**Comment by tstuefe** [comment]
Oh, I don't even have the first one... @shipilev, any more suggestions?

(this is not terribly important, no pressure)

File: test/hotspot/gtest/runtime/test_os.cpp
Comment (dholmes-ora) [inline test/hotspot/gtest/runtime/test_os.cpp:37]: Is this a leftover?

**Review by dholmes-ora — CHANGES_REQUESTED** [review]
Seems reasonable. Just one (hopefully) minor thing below.

Thanks.

File: test/hotspot/gtest/runtime/test_os.cpp
Comment (tstuefe) [reply test/hotspot/gtest/runtime/test_os.cpp]: Oops, thank you. That slipped by me, it's from an earlier experiment.

**Review by shipilev — APPROVED** [review]
Looks good.

It's a pity that we would read full `intptr_t` for every unit, which would mean that reading byte-unit block of N bytes would issue 8*N SafeFetch reads. But I guess it is fine, given the old code did `SafeFetch32` anyway.

That looks like fixing another implicit bug, if we are reading the unaligned 64-bit location at readable boundary: `is_readable_pointer` would read 32-bits and return `true` for first 4 bytes of the location, and then we go in and read 64 bits with `print_hex_readable_pointer` with `unitsize = 8`. Which might break if the last 4 bytes of that read are actually outside the range.

**Comment by tstuefe** [comment]
> Looks good.
> 
> It's a pity that we would read full `intptr_t` for every unit, which would mean that reading byte-unit block of N bytes would issue 8*N SafeFetch reads. But I guess it is fine, given the old code did `SafeFetch32` anyway.

I thought about this too but then refrained from implementing a "read intptr_t then print bytes" to keep the patch complexity low. 

I think another, simpler, patch would be worthwhile where we split print_hex_dump into two variants, one using SafeFetch and one just reading, and use the former one only if we are unsure about the memory content. E.g. at hs-err file dumping. There is no need to pay for SafeFetch when hexdumping e.g. for logging.

Maybe for a future RFE.

> 
> That looks like fixing another implicit bug, if we are reading the unaligned 64-bit location at readable boundary: `is_readable_pointer` would read 32-bits and return `true` for first 4 bytes of the location, and then we go in and read 64 bits with `print_hex_readable_pointer` with `unitsize = 8`. Which might break if the last 4 bytes of that read are actually outside the readable range.

Oh, I missed that. Nice :)

Thanks for the review!

**Comment by tstuefe** [comment]
@dholmes-ora are you okay with the patch in its current form?

**Review by dholmes-ora — APPROVED** [review]
Looks okay to me. Thanks

**Comment by tstuefe** [comment]
Thanks David!

/integrate

**Pull request closed without merging**
